jax_distribution/utils/logger_tools.py

```diff
@@ -0,0 +1,108 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import logging
+from collections import defaultdict
+from copy import deepcopy
+from typing import Any, Dict, List, Optional
+
+import numpy as np
+from neptune import new as neptune
+
+
+class Logger:
+    """Logger class for logging to tensorboard, sacred, and neptune."""
+
+    def __init__(self, console_logger: logging.Logger) -> None:
+        """Initialise the logger."""
+        self.console_logger = console_logger
+
+        self.use_tb = False
+        self.use_sacred = False
+        self.use_neptune = False
+
+        self.stats = defaultdict(lambda: [])
+
+    def setup_tb(self, directory_name: str) -> None:
+        """Set up tensorboard logging."""
+        # Import here so it doesn't have to be installed if you don't use it
+        from tensorboard_logger import configure, log_value
+
+        configure(directory_name)
+        self.tb_logger = log_value
+        self.use_tb = True
+
+    def setup_sacred(self, sacred_run_dict: Dict) -> None:
+        """Set up sacred logging."""
+        self._run_obj = sacred_run_dict
+        self.sacred_info = sacred_run_dict.info
+        self.use_sacred = True
+
+    def setup_neptune(self, neptune_run_dict: Dict) -> None:
+        pass
+
+    def log_stat(self, key: str, value: float, t: int, to_sacred: bool = True) -> None:
+        """Log a single stat."""
+        self.stats[key].append((t, value))
+
+        if self.use_tb:
+            self.tb_logger(key, value, t)
+
+        if self.use_sacred and to_sacred:
+            if key in self.sacred_info:
+                self.sacred_info["{}_T".format(key)].append(t)
+                self.sacred_info[key].append(value)
+            else:
+                self.sacred_info["{}_T".format(key)] = [t]
+                self.sacred_info[key] = [value]
+
+            self._run_obj.log_scalar(key, value, t)
+
+
+def get_logger() -> logging.Logger:
+    """Set up a custom logger."""
+    logger = logging.getLogger()
+    logger.handlers = []
+    ch = logging.StreamHandler()
+    formatter = logging.Formatter(
+        "[%(levelname)s %(asctime)s] %(name)s %(message)s", "%H:%M:%S"
+    )
+    ch.setFormatter(formatter)
+    logger.addHandler(ch)
+    # Set to info to suppress debug outputs.
+    logger.setLevel("INFO")
+
+    return logger
+
+
+def recursive_dict_update(d: Dict, u: Dict) -> Dict:
+    """Recursively update a dictionary."""
+    for k, v in u.items():
+        if isinstance(v, collections.Mapping):
+            d[k] = recursive_dict_update(d.get(k, {}), v)
+        else:
+            d[k] = v
+    return d
+
+
+def config_copy(config: Dict) -> Dict:
```

Comment:
This is because the _config from sacred is set to be read-only however we make changes to config while running the system to get different needed information shared between the different functions thanks to config

Reply:
if you have another option happy to give it a try

Reply:
We will add it as todo in the future